src/main/java/org/reactome/addlinks/dataretrieval/ensembl/EnsemblBioMartRetriever.java

```diff
@@ -0,0 +1,247 @@
+        final int MAX_QUERY_RETRIES = 5;
+        // Create connection to BioMart URL for each species, retrieving a list of microarray probe types, if available.
+        URL biomartUrlWithSpecies = new URL(queryString);
+        HttpURLConnection biomartConnection = (HttpURLConnection) biomartUrlWithSpecies.openConnection();
+        if (biomartConnection.getResponseCode() == HttpStatus.SC_OK) {
+            Set<String> biomartResponseLines = new HashSet<>();
+            BufferedReader br = new BufferedReader(new InputStreamReader(biomartConnection.getInputStream()));
+            String line;
+            while ((line = br.readLine()) != null) {
+                // BioMart still responds with a 200, even if no data exists. For now, we handle it by
+                // checking the returned content for the string 'ERROR'. It will retry up to 5 times, with a 10 second delay.
+                if (line.contains("ERROR") ) {
+                    retryCount++;
+                    if (retryCount < MAX_QUERY_RETRIES) {
+                        return retryQuery(queryString, retryCount);
+                    }
+                    // The data does not exist in BioMart for a few species. Frog (X. tropicalis) has UniProt data but not Microarray data;
+                    // Yeast (S. cerevisiae) has Uniprot-SwissProt data but not UniProt-TrEMBL data; Yeast (S. pombe), P. falciparum
+                    // and D. discoideum don't have UniProt or Microarray data, all at time of writing (January 2020).
+                    throw new BioMartQueryException(line +
+                            "\nThis can happen without issue for certain species (X. tropicalis, S. pombe, S. cerevisiae, P. falciparum) " +
+                            "because the data doesn't exist in BioMart");
+                } else {
+                    biomartResponseLines.add(line);
+                }
+            }
+            return biomartResponseLines;
+
+        } else {
+            retryCount++;
+            if (retryCount < MAX_QUERY_RETRIES) {
+                return retryQuery(queryString, retryCount);
+            }
+            throw new HttpException(
+                    "Unable to connect to BioMart (" +
+                            biomartConnection.getResponseCode() + ": " + biomartConnection.getResponseMessage() +
+                            ") with URL: " + queryString
+            );
+        }
+    }
+
+    // Recursive method that retries the BioMart query.
+    private Set<String> retryQuery(String queryString, int retryCount) throws InterruptedException, IOException, HttpException, BioMartQueryException {
+        final long QUERY_SLEEP_DURATION = Duration.ofSeconds(5).toMillis();
+
+        Thread.sleep(QUERY_SLEEP_DURATION);
+        logger.warn("BioMart query failed. Trying again...");
+        return queryBioMart(queryString, retryCount);
```

Comment:
`queryBioMart` calls `retryQuery` and `retryQuery` calls `queryBioMart`?? This is confusing, and it looks unnecessary. Why not just have all of the logic in one function?

Reply:
Ironically, I did this thinking it would make the recursion of the function more obvious/understandable. Since that seems to _not_ be the case, maybe I'll roll it into one function

Reply:
@SolomonShorser-OICR The other reason for having the secondary method was that there are two areas in 'queryBiomart' that call 'retryQuery', so I created the helper to consolidate it

Reply:
@cookersjs Ah. If there's different things that call `retryQuery` then it makes a little more sense.